File: lint.py
#!/usr/bin/env python3

import argparse
import collections
from pathlib import Path
from typing import Optional
import colorama
import fnmatch
import functools
import os.path
import re
import sys
import os
import subprocess

try:
    from PIL import Image

    PILLOW_INSTALLED = True
except ImportError:
    print("Pillow could not be imported - will not run image checks")
    print("Install pillow with `pip3 install pillow`")
    PILLOW_INSTALLED = False


class AnsiFore:
    KEEP = ""
    BLACK = "\033[30m"
    RED = "\033[31m"
    GREEN = "\033[32m"
    YELLOW = "\033[33m"
    BLUE = "\033[34m"
    MAGENTA = "\033[35m"
    CYAN = "\033[36m"
    WHITE = "\033[37m"
    RESET = "\033[39m"

    BOLD_BLACK = "\033[1;30m"
    BOLD_RED = "\033[1;31m"
    BOLD_GREEN = "\033[1;32m"
    BOLD_YELLOW = "\033[1;33m"
    BOLD_BLUE = "\033[1;34m"
    BOLD_MAGENTA = "\033[1;35m"
    BOLD_CYAN = "\033[1;36m"
    BOLD_WHITE = "\033[1;37m"
    BOLD_RESET = "\033[1;39m"


class AnsiStyle:
    BRIGHT = "\033[1m"
    BOLD = "\033[1m"
    DIM = "\033[2m"
    THIN = "\033[2m"
    NORMAL = "\033[22m"
    RESET_ALL = "\033[0m"


Fore = AnsiFore()
Style = AnsiStyle()


def print_error_for_file(file: str, body: Optional[str]):
    print(f"{AnsiFore.GREEN}### File {AnsiStyle.BRIGHT}{file}{AnsiStyle.RESET_ALL}")
    print()
    if body is not None:
        print(body)
        print()


def git_ls_files(patterns=None):
    command = ["git", "ls-files", "-s"]
    if patterns is not None:
        command.extend(patterns)
    proc = subprocess.Popen(command, stdout=subprocess.PIPE)
    output, _ = proc.communicate()
    lines = [x.split() for x in output.decode("utf-8").splitlines()]
    return {s[3].strip(): int(s[0]) for s in lines}


def find_all(a_str, sub):
    if not a_str.find(sub):
        # Optimization: If str is not in whole text, then do not try
        # on each line
        return
    for i, line in enumerate(a_str.split("\n")):
        column = 0
        while True:
            column = line.find(sub, column)
            if column == -1:
                break
            yield i, column
            column += len(sub)


colorama.init()

parser = argparse.ArgumentParser()
parser.add_argument(
    "files", nargs="*", default=[], help="files to be processed (regex on path)"
)
args = parser.parse_args()

EXECUTABLE_BIT = git_ls_files()
files = list(EXECUTABLE_BIT.keys())
# Match against re
file_name_re = re.compile("|".join(args.files))
files = [p for p in files if file_name_re.search(p)]

files.sort()

file_types = (
    ".cfg",
    ".css",
    ".gif",
    ".h",
    ".html",
    ".ico",
    ".jpg",
    ".js",
    ".json",
    ".md",
    ".png",
    ".py",
    ".svg",
    ".toml",
    ".txt",
    ".webmanifest",
    ".xml",
    ".yaml",
    ".yml",
    "",
)
docs_types = [".md"]
image_types = [".webp", ".jpg", ".ico", ".png", ".svg", ".gif"]

LINT_FILE_CHECKS = []
LINT_CONTENT_CHECKS = []
LINT_POST_CHECKS = []


def run_check(lint_obj, fname, *args):
    include = lint_obj["include"]
    exclude = lint_obj["exclude"]
    func = lint_obj["func"]
    if include is not None:
        for incl in include:
            if fnmatch.fnmatch(fname, incl):
                break
        else:
            return None
    for excl in exclude:
        if fnmatch.fnmatch(fname, excl):
            return None
    return func(*args)


def run_checks(lints, fname, *args):
    for lint in lints:
        try:
            add_errors(fname, run_check(lint, fname, *args))
        except Exception:
            print(f"Check {lint['func'].__name__} on file {fname} failed:")
            raise


def _add_check(checks, func, include=None, exclude=None):
    checks.append(
        {
            "include": include,
            "exclude": exclude or [],
            "func": func,
        }
    )


def lint_file_check(**kwargs):
    def decorator(func):
        _add_check(LINT_FILE_CHECKS, func, **kwargs)
        return func

    return decorator


def lint_content_check(**kwargs):
    def decorator(func):
        _add_check(LINT_CONTENT_CHECKS, func, **kwargs)
        return func

    return decorator


def lint_post_check(func):
    _add_check(LINT_POST_CHECKS, func)
    return func


def lint_re_check(regex, **kwargs):
    flags = kwargs.pop("flags", re.MULTILINE)
    prog = re.compile(regex, flags)
    decor = lint_content_check(**kwargs)

    def decorator(func):
        @functools.wraps(func)
        def new_func(fname, content):
            errors = []
            for match in prog.finditer(content):
                if "NOLINT" in match.group(0):
                    continue
                lineno = content.count("\n", 0, match.start()) + 1
                substr = content[: match.start()]
                col = len(substr) - substr.rfind("\n")
                err = func(fname, match)
                if err is None:
                    continue
                errors.append((lineno, col + 1, err))
            return errors

        return decor(new_func)

    return decorator


def lint_content_find_check(find, only_first=False, **kwargs):
    decor = lint_content_check(**kwargs)

    def decorator(func):
        @functools.wraps(func)
        def new_func(fname, content):
            find_ = find
            if callable(find):
                find_ = find(fname, content)
            errors = []
            for line, col in find_all(content, find_):
                err = func(fname)
                errors.append((line + 1, col + 1, err))
                if only_first:
                    break
            return errors

        return decor(new_func)

    return decorator


@lint_file_check(exclude=[f"*{f}" for f in file_types])
def lint_ext_check(fname: str, stat: os.stat_result):
    return (
        "This file extension is not a registered file type. If this is an error, please "
        "update the script/ci-custom.py script."
    )


@lint_file_check(exclude=["script/*", ".devcontainer/*", "lint.py"])
def lint_executable_bit(fname: str, stat: os.stat_result):
    ex = EXECUTABLE_BIT[fname]
    if ex != 100644:
        return (
            "File has invalid executable bit {}. If running from a windows machine please "
            "see disabling executable bit in git.".format(ex)
        )
    return None


@lint_file_check(
    include=[f"images/*{ext}" for ext in image_types], exclude=["images/hero.png"]
)
def lint_index_images_size(fname: str, stat: os.stat_result):
    if stat.st_size > 40 * 1024:
        return (
            "Image is too large. The files in the images/ folder are displayed on esphome's "
            "front page and thus should be small (no more than 300x300px, and <40kb). "
            "Use a tool like https://compress-or-die.com/ to reduce the image size. "
            f"Size of file: {stat.st_size / 1024:.0f}kb"
        )
    return None


@lint_file_check(include=[f"*{ext}" for ext in image_types])
def lint_all_images_size(fname: str, stat: os.stat_result):
    if stat.st_size > 1024 * 1024:
        return (
            "Image is too large. Images in ESPHome's codebase should be 1MB in size max. "
            "Use a tool like https://compress-or-die.com/ to reduce the image size. "
            f"Size of file: {stat.st_size / 1024:.0f}kb"
        )
    return None


if PILLOW_INSTALLED:

    @lint_file_check(
        include=["images/*.jpg", "images/*.png"], exclude=["images/hero.png"]
    )
    def lint_index_images_dimensions(fname: str, stat: os.stat_result):
        img = Image.open(fname)
        if img.width > 300 or img.height > 300:
            return (
                "Image has too large dimensions. The images in the images/ folder are displayed on "
                "ESPHome's main page, so need to be lightweight. We allow a max of 300x300 for images on this page. "
                "Use a tool like https://compress-or-die.com/ to reduce the image size. "
                f"Dimensions of this image: {img.width}x{img.height}"
            )
        return None


@lint_content_find_check(
    "\t",
    only_first=True,
    exclude=[
        "Makefile",
    ],
)
def lint_tabs(fname):
    return "File contains tab character. Please convert tabs to spaces."


@lint_content_find_check("\r", only_first=True)
def lint_newline(fname):
    return "File contains Windows newline. Please set your editor to Unix newline mode."


@lint_content_check(exclude=["*.svg", "runtime.txt", "_static/*"])
def lint_end_newline(fname, content):
    if content and not content.endswith("\n"):
        return "File does not end with a newline, please add an empty line at the end of the file."
    return None


@lint_re_check(
    r"\[([^\]]+)\]\((https://esphome\.io/[^)]+)\)",
    include=["*.md"],
)
def lint_esphome_io_link(fname, match):
    link_text = match.group(1)
    full_url = match.group(2)
    return (
        f"Markdown link to esphome.io should use relative path. "
        f"Change [{link_text}]({full_url}) to use a relative URL"
    )


# Build cache of all anchors in the documentation
ANCHOR_CACHE = None


def build_anchor_cache():
    """Build a cache of all anchors (headings and shortcodes) in markdown files."""
    global ANCHOR_CACHE
    if ANCHOR_CACHE is not None:
        return ANCHOR_CACHE

    ANCHOR_CACHE = {}
    content_dir = Path("content")
    if not content_dir.exists():
        return ANCHOR_CACHE

    def generate_slug(text):
        """Generate Hugo-compatible slug from heading text."""
        slug = re.sub(r"[^\w\s-]", "", text).lower()
        slug = re.sub(r"[-\s]+", "-", slug).strip("-")
        return slug

    for md_file in content_dir.rglob("*.md"):
        rel_path = md_file.relative_to(content_dir)
        # Convert file path to page path
        if md_file.name == "_index.md":
            if str(rel_path) == "_index.md":
                page_path = ""  # Root index
            else:
                page_path = str(rel_path.parent)
        else:
            page_path = str(rel_path.with_suffix(""))

        page_path = page_path.replace("\\", "/")
        anchors = set()

        try:
            with open(md_file, "r", encoding="utf-8") as f:
                content = f.read()
                lines = content.split("\n")

            for line in lines:
                # Match headings
                heading_match = re.match(r"^(#{1,6})\s+(.*)", line)
                if heading_match:
                    heading_text = heading_match.group(2).strip()
                    anchor_id = generate_slug(heading_text)
                    anchors.add(anchor_id)

                # Match anchor shortcodes
                shortcode_match = re.search(r'\{\{<\s*anchor\s+"([^\s>]+)"\s*>}}', line)
                if shortcode_match:
                    anchor_id = shortcode_match.group(1)
                    anchors.add(anchor_id)

            ANCHOR_CACHE[page_path] = anchors
        except Exception:
            # Skip files that can't be read
            pass

    return ANCHOR_CACHE


@lint_content_check(include=["*.md"])
def lint_internal_links(fname, content):
    """Validate internal markdown links."""
    errors = []
    build_anchor_cache()

    # Match markdown links: [text](url)
    # Capture groups: text and URL
    link_pattern = re.compile(r"\[([^\]]+)\]\(([^)]+)\)")

    for match in link_pattern.finditer(content):
        link_text = match.group(1)
        link_url = match.group(2)
        lineno = content.count("\n", 0, match.start()) + 1
        col = match.start() - content.rfind("\n", 0, match.start())

        # Skip external links (http://, https://, mailto:, etc.)
        if re.match(r"^[a-zA-Z][a-zA-Z0-9+.-]*:", link_url):
            continue

        # Check anchor-only links (these used to be resolved by the old system)
        if link_url.startswith("#"):
            anchor_id = link_url.lstrip("#")

            # Get current page path (only for files in content/)
            current_file = Path(fname)
            try:
                if current_file.name == "_index.md":
                    if str(current_file.parent) == "content":
                        current_page = ""
                    else:
                        current_page = str(current_file.parent.relative_to("content"))
                else:
                    current_page = str(current_file.relative_to("content").with_suffix(""))

                current_page = current_page.replace("\\", "/")

                # Check if anchor exists on current page
                if current_page in ANCHOR_CACHE:
                    if anchor_id not in ANCHOR_CACHE[current_page]:
                        # Anchor doesn't exist on current page - likely a broken cross-page link
                        errors.append(
                            (
                                lineno,
                                col,
                                f"Anchor-only link '#{anchor_id}' not found on current page. "
                                f"If this should link to another page, use the full path like "
                                f"[{link_text}](/path/to/page#{anchor_id})",
                            )
                        )
            except ValueError:
                # File is not in content/ directory, skip anchor validation
                pass
            continue

        # Skip relative links to static assets (images, etc.)
        # These don't start with / and have file extensions
        if not link_url.startswith("/") and re.search(r"\.(png|jpg|jpeg|gif|svg|webp|pdf|zip)$", link_url, re.IGNORECASE):
            continue

        # Skip links that look like code/lambda parameters (contain spaces, parentheses, etc.)
        if " " in link_url or "(" in link_url or ")" in link_url:
            continue

        # Skip relative links without leading slash (not documentation pages)
        # unless they're in the components directory
        if not link_url.startswith("/"):
            # Allow relative component links
            if not ("components/" in fname and not re.search(r"\.", link_url)):
                continue

        # Parse internal link
        if "#" in link_url:
            path_part, anchor_part = link_url.split("#", 1)
        else:
            path_part = link_url
            anchor_part = None

        # Validate path - must start with /
        if not path_part.startswith("/"):
            continue

        # Remove leading slash
        clean_path = path_part.lstrip("/")

        # Skip paths with query strings or fragments that look like URLs
        if "?" in clean_path or ".html" in clean_path:
            continue

        # Check if page exists
        if clean_path not in ANCHOR_CACHE:
            # Try with _index
            if clean_path and not clean_path.endswith("/"):
                clean_path_index = clean_path
            else:
                clean_path_index = clean_path.rstrip("/")

            if clean_path_index not in ANCHOR_CACHE:
                errors.append(
                    (
                        lineno,
                        col,
                        f"Internal link references non-existent page: '{path_part}' in link [{link_text}]({link_url})",
                    )
                )
                continue

        # Validate anchor if present
        if anchor_part:
            target_page = clean_path
            if target_page not in ANCHOR_CACHE:
                target_page = clean_path.rstrip("/")

            if target_page in ANCHOR_CACHE:
                if anchor_part not in ANCHOR_CACHE[target_page]:
                    errors.append(
                        (
                            lineno,
                            col,
                            f"Internal link references non-existent anchor: '#{anchor_part}' on page '{path_part}' in link [{link_text}]({link_url})",
                        )
                    )

    return errors


def highlight(s):
    return f"\033[36m{s}\033[0m"


errors = collections.defaultdict(list)


def add_errors(fname, errs):
    if not isinstance(errs, list):
        errs = [errs]
    for err in errs:
        if err is None:
            continue
        try:
            lineno, col, msg = err
        except ValueError:
            lineno = 1
            col = 1
            msg = err
        if not isinstance(msg, str):
            raise ValueError("Error is not instance of string!")
        if not isinstance(lineno, int):
            raise ValueError("Line number is not an int!")
        if not isinstance(col, int):
            raise ValueError("Column number is not an int!")
        errors[fname].append((lineno, col, msg))


for fname in files:
    p = Path(fname)
    if not p.is_file():
        # file deleted but in git index
        continue
    run_checks(LINT_FILE_CHECKS, fname, fname, p.stat())
    if p.suffix in image_types:
        continue
    try:
        with open(fname, "r") as f_handle:
            content = f_handle.read()
    except UnicodeDecodeError:
        add_errors(
            fname,
            "File is not readable as UTF-8. Please set your editor to UTF-8 mode.",
        )
        continue
    run_checks(LINT_CONTENT_CHECKS, fname, fname, content)

run_checks(LINT_POST_CHECKS, "POST")

for f, errs in sorted(errors.items()):
    err_str = (
        f"{AnsiStyle.BOLD}{f}:{lineno}:{col}:{AnsiStyle.RESET_ALL} "
        f"{AnsiStyle.BOLD}{AnsiFore.RED}{'lint:'}{AnsiStyle.RESET_ALL} {msg}\n"
        for lineno, col, msg in errs
    )
    print_error_for_file(f, "\n".join(err_str))

sys.exit(len(errors))

File: lint.mjs
#!/usr/bin/env node

import { readFile, stat, readdir } from 'fs/promises';
import { join, relative, basename, dirname, extname } from 'path';
import { execSync } from 'child_process';
import { fileURLToPath } from 'url';

const __filename = fileURLToPath(import.meta.url);
const __dirname = dirname(__filename);

// ANSI color codes
const colors = {
  reset: '\x1b[0m',
  bright: '\x1b[1m',
  dim: '\x1b[2m',
  red: '\x1b[31m',
  green: '\x1b[32m',
  yellow: '\x1b[33m',
  cyan: '\x1b[36m',
  boldRed: '\x1b[1;31m',
  boldGreen: '\x1b[1;32m',
};

// Folders to ignore
const ignoreFolders = [
  'pagefind/',
  'node_modules/',
  'dist/',
  '.astro/',
];

// File types
const fileTypes = [
  '.cfg', '.css', '.gif', '.h', '.html', '.ico', '.jpg', '.js', '.json',
  '.md', '.mdx', '.png', '.py', '.svg', '.toml', '.txt', '.webmanifest',
  '.xml', '.yaml', '.yml', '.mjs', '.ts', '.tsx', '.astro', '.sh', '.webp',
  '' // empty string for files without extension (like .gitignore)
];
const imageTypes = ['.webp', '.jpg', '.ico', '.png', '.svg', '.gif'];

// Store errors
const errors = new Map();

function addError(fname, lineno, col, msg) {
  if (!errors.has(fname)) {
    errors.set(fname, []);
  }
  errors.get(fname).push({ lineno, col, msg });
}

// Get files from git
function getGitFiles() {
  try {
    const output = execSync('git ls-files -s', { encoding: 'utf-8' });
    const files = new Map();

    for (const line of output.trim().split('\n')) {
      const parts = line.split(/\s+/);
      if (parts.length >= 4) {
        const mode = parseInt(parts[0]);
        const path = parts.slice(3).join(' ');
        files.set(path, mode);
      }
    }

    return files;
  } catch (error) {
    console.error('Error getting git files:', error.message);
    return new Map();
  }
}

// Find all occurrences of a substring
function* findAll(content, substring) {
  const lines = content.split('\n');
  for (let i = 0; i < lines.length; i++) {
    let col = 0;
    while (true) {
      col = lines[i].indexOf(substring, col);
      if (col === -1) break;
      yield { line: i + 1, col: col + 1 };
      col += substring.length;
    }
  }
}

// File checks
async function checkFileExtension(fname) {
  const ext = extname(fname);
  if (!fileTypes.includes(ext)) {
    addError(fname, 1, 1,
      'This file extension is not a registered file type. If this is an error, please update lint.mjs.');
  }
}

async function checkExecutableBit(fname, gitMode) {
  const exclude = ['script/', '.devcontainer/', 'lint.py', 'lint.mjs'];
  if (exclude.some(ex => fname.startsWith(ex) || fname === ex)) {
    return;
  }

  if (gitMode !== 100644) {
    addError(fname, 1, 1,
      `File has invalid executable bit ${gitMode}. If running from a windows machine please see disabling executable bit in git.`);
  }
}

async function checkImageSize(fname, fileStat) {
  const ext = extname(fname);
  if (!imageTypes.includes(ext)) return;

  if (fileStat.size > 1024 * 1024) {
    const sizeKb = Math.floor(fileStat.size / 1024);
    addError(fname, 1, 1,
      `Image is too large. Images should be 1MB max. Use https://compress-or-die.com/ to reduce size. Current size: ${sizeKb}kb`);
  }
}

// Content checks
function checkTabs(fname, content) {
  if (fname === 'Makefile') return;

  for (const { line, col } of findAll(content, '\t')) {
    addError(fname, line, col, 'File contains tab character. Please convert tabs to spaces.');
    return; // Only report first occurrence
  }
}

function checkNewlines(fname, content) {
  for (const { line, col } of findAll(content, '\r')) {
    addError(fname, line, col, 'File contains Windows newline. Please set your editor to Unix newline mode.');
    return; // Only report first occurrence
  }
}

function checkEndNewline(fname, content) {
  const exclude = ['.svg', 'runtime.txt'];
  if (exclude.some(ex => fname.endsWith(ex)) || fname.includes('_static/')) {
    return;
  }

  if (content && !content.endsWith('\n')) {
    addError(fname, 1, 1, 'File does not end with a newline, please add an empty line at the end of the file.');
  }
}

function checkEsphomeLinks(fname, content) {
  if (!fname.endsWith('.md') && !fname.endsWith('.mdx')) return;

  const linkPattern = /\[([^\]]+)\]\((https:\/\/esphome\.io\/[^)]+)\)/g;
  let match;

  while ((match = linkPattern.exec(content)) !== null) {
    const lineno = content.substring(0, match.index).split('\n').length;
    const col = match.index - content.lastIndexOf('\n', match.index);
    const linkText = match[1];
    const fullUrl = match[2];

    addError(fname, lineno, col,
      `Markdown link to esphome.io should use relative path. Change [${linkText}](${fullUrl}) to use a relative URL`);
  }
}

// Build anchor cache for link validation
async function buildAnchorCache() {
  const cache = new Map();
  const contentDir = join(__dirname, 'src/content/docs');

  try {
    await stat(contentDir);
  } catch {
    return cache;
  }

  function generateSlug(text) {
    return text
      .toLowerCase()
      .replace(/[^\w\s-]/g, '')
      .replace(/[-\s]+/g, '-')
      .replace(/^-+|-+$/g, '');
  }

  async function processDir(dir, basePath = '') {
    const entries = await readdir(dir, { withFileTypes: true });

    for (const entry of entries) {
      const fullPath = join(dir, entry.name);
      const relPath = basePath ? join(basePath, entry.name) : entry.name;

      if (entry.isDirectory()) {
        await processDir(fullPath, relPath);
      } else if (entry.name.endsWith('.md') || entry.name.endsWith('.mdx')) {
        let pagePath;
        if (entry.name === 'index.md' || entry.name === 'index.mdx') {
          pagePath = basePath || '';
        } else {
          pagePath = relPath.replace(/\.mdx?$/, '');
        }

        pagePath = pagePath.replace(/\\/g, '/');

        const content = await readFile(fullPath, 'utf-8');
        const anchors = new Set();

        for (const line of content.split('\n')) {
          // Match headings
          const headingMatch = line.match(/^(#{1,6})\s+(.*)/);
          if (headingMatch) {
            const headingText = headingMatch[2].trim();
            const anchorId = generateSlug(headingText);
            anchors.add(anchorId);
          }

          // Match anchor shortcodes (old Hugo style)
          const shortcodeMatch = line.match(/\{\{<\s*anchor\s+"([^\s>]+)"\s*>}}/);
          if (shortcodeMatch) {
            anchors.add(shortcodeMatch[1]);
          }

          // Match HTML anchor spans
          const spanMatch = line.match(/<span id="([^"]+)"><\/span>/);
          if (spanMatch) {
            anchors.add(spanMatch[1]);
          }
        }

        cache.set(pagePath, anchors);
      }
    }
  }

  await processDir(contentDir);
  return cache;
}

async function checkInternalLinks(fname, content, anchorCache) {
  if (!fname.endsWith('.md') && !fname.endsWith('.mdx')) return;

  // Skip template and documentation files with example links
  const ignoreFiles = [
    '.claude/instructions.md',
    'script/release_notes_template.md',
  ];
  if (ignoreFiles.includes(fname)) return;

  const linkPattern = /\[([^\]]+)\]\(([^)]+)\)/g;
  let match;

  while ((match = linkPattern.exec(content)) !== null) {
    const linkText = match[1];
    const linkUrl = match[2];
    const lineno = content.substring(0, match.index).split('\n').length;
    const col = match.index - content.lastIndexOf('\n', match.index);

    // Skip external links
    if (/^[a-zA-Z][a-zA-Z0-9+.-]*:/.test(linkUrl)) {
      continue;
    }

    // Check anchor-only links
    if (linkUrl.startsWith('#')) {
      const anchorId = linkUrl.substring(1);

      // Get current page path
      let currentPage;
      try {
        if (fname.includes('src/content/docs/')) {
          const relPath = fname.split('src/content/docs/')[1];
          if (basename(relPath) === 'index.md' || basename(relPath) === 'index.mdx') {
            currentPage = dirname(relPath) === '.' ? '' : dirname(relPath);
          } else {
            currentPage = relPath.replace(/\.mdx?$/, '');
          }
          currentPage = currentPage.replace(/\\/g, '/');

          if (anchorCache.has(currentPage)) {
            if (!anchorCache.get(currentPage).has(anchorId)) {
              addError(fname, lineno, col,
                `Anchor-only link '#${anchorId}' not found on current page. If this should link to another page, use the full path like [${linkText}](/path/to/page#${anchorId})`);
            }
          }
        }
      } catch (e) {
        // Skip if not in content directory
      }
      continue;
    }

    // Skip relative links to static assets
    if (!linkUrl.startsWith('/') && /\.(png|jpg|jpeg|gif|svg|webp|pdf|zip)$/i.test(linkUrl)) {
      continue;
    }

    // Skip absolute links to static assets
    if (linkUrl.startsWith('/images/') && /\.(png|jpg|jpeg|gif|svg|webp|pdf|zip)$/i.test(linkUrl)) {
      continue;
    }

    // Skip links with spaces or parentheses (likely code)
    if (linkUrl.includes(' ') || linkUrl.includes('(') || linkUrl.includes(')')) {
      continue;
    }

    // Skip relative links without leading slash
    if (!linkUrl.startsWith('/')) {
      continue;
    }

    // Parse internal link
    let pathPart, anchorPart;
    if (linkUrl.includes('#')) {
      [pathPart, anchorPart] = linkUrl.split('#', 2);
    } else {
      pathPart = linkUrl;
      anchorPart = null;
    }

    // Skip paths with query strings or .html
    if (pathPart.includes('?') || pathPart.includes('.html')) {
      continue;
    }

    // Remove leading slash and trailing slash
    const cleanPath = pathPart.replace(/^\/+/, '').replace(/\/+$/, '');

    // Check if page exists
    // Try both with and without '/index' suffix since both are valid
    const pathsToCheck = [cleanPath];
    if (cleanPath.endsWith('/index')) {
      pathsToCheck.push(cleanPath.replace(/\/index$/, ''));
    } else if (cleanPath !== '') {
      pathsToCheck.push(cleanPath + '/index');
    }

    const pageExists = pathsToCheck.some(p => anchorCache.has(p) || p === '');
    if (!pageExists && cleanPath !== '') {
      addError(fname, lineno, col,
        `Internal link references non-existent page: '${pathPart}' in link [${linkText}](${linkUrl})`);
      continue;
    }

    // Validate anchor if present
    if (anchorPart && cleanPath) {
      // Find the actual page path in cache (might be with or without /index)
      let targetPage = cleanPath;
      if (!anchorCache.has(targetPage)) {
        if (cleanPath.endsWith('/index')) {
          targetPage = cleanPath.replace(/\/index$/, '');
        } else {
          const withIndex = cleanPath + '/index';
          if (anchorCache.has(withIndex)) {
            targetPage = withIndex;
          }
        }
      }

      if (anchorCache.has(targetPage)) {
        if (!anchorCache.get(targetPage).has(anchorPart)) {
          addError(fname, lineno, col,
            `Internal link references non-existent anchor: '#${anchorPart}' on page '${pathPart}' in link [${linkText}](${linkUrl})`);
        }
      }
    }
  }
}

function checkAutomationHeadings(fname, content) {
  if (!fname.startsWith('src/content/docs/components/')) return;
  if (!fname.endsWith('.md') && !fname.endsWith('.mdx')) return;

  // Skip the main actions page which documents core actions (delay, if, lambda, etc.)
  if (fname.includes('automations/')) return;

  const lines = content.split('\n');
  for (let i = 0; i < lines.length; i++) {
    const line = lines[i];
    const lineno = i + 1;

    // Only look at heading lines
    if (!line.match(/^#{2,4}\s/)) continue;

    // Check 1: Backticked `domain.name` without Action/Condition/Trigger suffix
    // e.g. ### `wireguard.enabled`
    if (line.match(/^#{2,4}\s+`[a-z_][a-z0-9_.]*\.[a-z_][a-z0-9_]*`\s*$/)) {
      addError(fname, lineno, 1,
        'Heading has backticked automation name but is missing Action/Condition/Trigger suffix. ' +
        'Add " Action", " Condition", or " Trigger" after the closing backtick.');
    }

    // Check 2: Action/Condition with backticked name missing domain prefix (no dot)
    // e.g. ### `arm_away` Action
    const noDotMatch = line.match(/^#{2,4}\s+`([a-z_][a-z0-9_]*)`\s+(?:Action|Condition)s?\s*$/i);
    if (noDotMatch) {
      const name = noDotMatch[1];
      // Exclude core actions/conditions documented on actions.mdx
      const coreNames = [
        'delay', 'if', 'lambda', 'repeat', 'wait_until', 'while',
        'and', 'all', 'or', 'any', 'xor', 'not', 'for',
      ];
      if (!coreNames.includes(name) && !name.startsWith('on_')) {
        addError(fname, lineno, 1,
          `Heading "\`${name}\`" is missing the domain prefix. ` +
          'Use the format: `domain.name` Action/Condition (e.g. `switch.toggle` Action).');
      }
    }

    // Check 3: Bold **Action** or **Condition** suffix
    // e.g. ### `remote_transmitter.transmit_nec` **Action**
    if (line.match(/^#{2,4}\s+`[^`]+`.*\*\*(?:Action|Condition)s?\*\*/)) {
      addError(fname, lineno, 1,
        'Action/Condition suffix should not be bold. Use plain text: " Action" or " Condition".');
    }

    // Check 4: Lowercase action/condition suffix (not matching standard capitalization)
    // e.g. ### `sprinkler.start_full_cycle` action
    if (line.match(/^#{2,4}\s+`[a-z_][a-z0-9_.]*\.[a-z_][a-z0-9_]*`\s+(?:action|condition)s?\s*$/)) {
      addError(fname, lineno, 1,
        'Action/Condition suffix should be capitalized. Use "Action" or "Condition" (not lowercase).');
    }
  }
}

// Main execution
async function main() {
  console.log(`${colors.cyan}Running ESPHome documentation linter...${colors.reset}\n`);

  const gitFiles = getGitFiles();
  const anchorCache = await buildAnchorCache();

  // Process files
  for (const [fname, gitMode] of gitFiles) {
    // Skip ignored folders
    if (ignoreFolders.some(folder => fname.startsWith(folder) || fname.includes(`/${folder}`))) {
      continue;
    }

    try {
      const fileStat = await stat(fname);

      if (!fileStat.isFile()) continue;

      // Run file checks
      await checkFileExtension(fname);
      await checkExecutableBit(fname, gitMode);
      await checkImageSize(fname, fileStat);

      // Skip binary files
      if (imageTypes.includes(extname(fname))) {
        continue;
      }

      // Read and check content
      let content;
      try {
        content = await readFile(fname, 'utf-8');
      } catch (error) {
        if (error.message.includes('invalid')) {
          addError(fname, 1, 1, 'File is not readable as UTF-8. Please set your editor to UTF-8 mode.');
          continue;
        }
        throw error;
      }

      // Run content checks
      checkTabs(fname, content);
      checkNewlines(fname, content);
      checkEndNewline(fname, content);
      checkEsphomeLinks(fname, content);
      checkAutomationHeadings(fname, content);
      await checkInternalLinks(fname, content, anchorCache);

    } catch (error) {
      if (error.code !== 'ENOENT') {
        console.error(`Error processing ${fname}:`, error.message);
      }
    }
  }

  // Print errors
  if (errors.size > 0) {
    for (const [fname, errs] of [...errors.entries()].sort()) {
      console.log(`${colors.boldGreen}### File ${colors.bright}${fname}${colors.reset}\n`);

      for (const { lineno, col, msg } of errs) {
        console.log(
          `${colors.bright}${fname}:${lineno}:${col}:${colors.reset} ` +
          `${colors.bright}${colors.boldRed}lint:${colors.reset} ${msg}`
        );
      }
      console.log();
    }

    console.log(`${colors.boldRed}Found ${errors.size} file(s) with errors${colors.reset}`);
    process.exit(errors.size);
  } else {
    console.log(`${colors.boldGreen}✓ All checks passed!${colors.reset}`);
    process.exit(0);
  }
}

main().catch(error => {
  console.error(`${colors.boldRed}Fatal error:${colors.reset}`, error);
  process.exit(1);
});
